### DISCUSSION TITLE
Cancelling a async_session with task.cancel() causes a connection to be closed

### DISCUSSION THREAD

**Discussion by Kurtoid** [sqlalchemy/sqlalchemy#7523] [Usage Questions]
I'm trying to search an in-memory sqllite+aiosqlite database in a keyboard event handler. When some text is received, I check if there is running search task and cancel it with task.cancel(), and create a new one.
This works fine, as long as I don't type too fast: If the task gets canceled during a `async with async_session() as session:` context, any further events cause a `Cannot operate on a closed database.` error. I'm not sure what the best practices are with asyncio and SQLAlchemy, so I'm 90% sure I'm just doing this incorrectly.

I'm using an in memory sqllite database that I connect to with
`engine = create_async_engine("sqlite+aiosqlite:///:memory:?cache=some_name', echo=False, future=True)`

My event handler looks like this:
```python
def search_text_changed(self, buffer):
    if self.search_task and not self.search_task.done():
        self.search_task.cancel()
    self.search_task = asyncio.create_task(
        self.search(buffer.text, self.current_field))
```
and  the search function looks like
```python
async def search(self, text: str, field: Field):
    try:
        results_list = []
        engine = self.app.engine
        async_session = sessionmaker(bind=engine, class_=AsyncSession)
        async with async_session() as session:
            async with session.begin():
                results_list = await session.execute(
                    select([table_object]).where(
                        table_object.c[field.name].like(f"%{text}%")))
                results_list = results_list.fetchall()
        for record in results_list:
            # processing happens here; trimmed for brevity
            await asyncio.sleep(0)
    except CancelledError:
        return
    except Exception as e:
        print_exc()
        raise e
```

I'm thinking I have to do something special when either calling `cancel` or in the `CancelledError` handler, but I'm not sure what. Any ideas how I should do this?

**Comment by zzzeek** [comment]
the connection pool, even in the case of sqlite memory database, has a default behavior that when the connection is returned to the pool, .rollback() is called.  so first Q is do you see warnings when you run this?  you should see a warning that looks like:

```
                        "The garbage collector is trying to clean up "
                        "connection %r. This feature is unsupported on async "
                        "dbapi, since no IO can be performed at this stage to "
                        "reset the connection. Please close out all "
                        "connections when they are no longer used, calling "
                        "``close()`` or using a context manager to "
                        "manage their lifetime."
```

now the next thing is that this "reset on return" behavior can be turned off, so that in theory, even for an async dialect, we can just stick the connection back in the pool and there'd be no issue, the [pool_reset_on_return](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=pool_reset_on_return#sqlalchemy.create_engine.params.pool_reset_on_return) parameter allwos this.  *however*, it is not implemented right now that the above check / warning / discard of the connection will be skipped if this flag is turned off because we have not had this use case.

however again, the whole idea of the above warning is that no matter what, we were assuming that especially under asyncio, correct end-user code should be at all times making sure connections are released and/or closed out cleanly, including for anticipated error condtiions.  at the same time you are using the context manager which is supposed to be handling that.  in your CancelledError handler, the Session's context manager has already exited so I dont know you have a proper opportunity to clean things up there.

from my end I am not yet clear on how CancelledErrors are supposed to be handled in asyncio, the Python docs are fairly vague on this and for example I'm not sure if it's legal to invoke awaitables inside the handler for CancelledError, which is what we would need for there to be correct connection cleanup.

still more details I remember is that I'm pretty sure we consider CancelledError to be a "disconnect" condition regardless; that is, the state of an asyncio connection can't be determined when a CancelledError is caught, especially for example MySQL connections, so in the default case we have no choice but to throw it away.

so what's special here, you're using a SQLite "memory" connection.  This is a wholly different thing, and a really unusual use case.   There's no "IO" with a SQLite memory database, and even with a SQLite file database, while there is IO, it's only file IO which is not non-blocking even under asyncio.   

If the issue is that the CancelledError has to just not result in an "is_disconnect" situation, the approach at https://docs.sqlalchemy.org/en/14/core/events.html#sqlalchemy.events.ConnectionEvents.handle_error can be used to provide this, by setting "context.is_disconnect = False" in the event handler.

This is totally a lot to be presenting here and it also seems clear there's a new use case here that's been unanticipated.  if you can work your code fragment into a runnable single-file/no-dependencies example I can try, I can pinpoint exactly why the connenction is getting disconnected and hopefully show how to use a handle_error event to stop it from happening.   We'd likely need to add this to the SQLite dialect documentation as an example.
